src/schemas.jl

```diff
@@ -254,35 +254,79 @@ accepted_field_type(::SchemaVersion, ::Type{Symbol}) = Union{Symbol,String}
 """
 find_violation(::Tables.Schema, sv::SchemaVersion) = throw(UnknownSchemaVersionError(sv))
 
 function _find_violation end
 
+"""
+    Legolas.find_violations(ts::Tables.Schema, sv::Legolas.SchemaVersion)
+
+Return vector of all schema violations. For required fields `f::F` of `sv`:
+
+- Define `A = Legolas.accepted_field_type(sv, F)`
+- If `f::T` is present in `ts`, ensure that `T <: A` or else append `f::Symbol => T::DataType` to output vector.
+- If `f` isn't present in `ts`, ensure that `Missing <: A` or else append `f::Symbol => missing::Missing` to output vector.
+
+Otherwise, return `Any[]`.
+
+See also: [`Legolas.find_violation`](@ref), [`Legolas.validate`](@ref), [`Legolas.complies_with`](@ref)
+"""
+find_violations(::Tables.Schema, sv::SchemaVersion) = throw(UnknownSchemaVersionError(sv))
+
+function _find_violations end
+
 """
     Legolas.validate(ts::Tables.Schema, sv::Legolas.SchemaVersion)
 
-Throws a descriptive `ArgumentError` if `!isnothing(find_violation(ts, sv))`,
+Throws a descriptive `ArgumentError` if `!isempty(find_violations(ts, sv))`,
 otherwise return `nothing`.
 
-See also: [`Legolas.find_violation`](@ref), [`Legolas.complies_with`](@ref)
+See also: [`Legolas.find_violations`](@ref), [`Legolas.find_violation`](@ref), [`Legolas.complies_with`](@ref)
 """
 function validate(ts::Tables.Schema, sv::SchemaVersion)
-    result = find_violation(ts, sv)
-    isnothing(result) && return nothing
-    field, violation = result
-    ismissing(violation) && throw(ArgumentError("could not find expected field `$field` in $ts"))
-    expected = getfield(required_fields(sv), field)
-    throw(ArgumentError("field `$field` has unexpected type; expected <:$expected, found $violation"))
+    results = find_violations(ts, sv)
+    isempty(results) && return nothing
+
+    field_err = Any[]
+    type_err = Any[]
+    for result in results
+        field, violation = result
+        if ismissing(violation)
+            push!(field_err, field)
+        else
+            expected = getfield(required_fields(sv), field)
+            push!(type_err, (field, expected, violation))
+        end
+    end
+    err = ""
+    if !isempty(field_err)
+        fields = "`" * join(field_err, "`, `") * "`"
+        err *= string("Could not find expected field", length(field_err) > 1 ? "s " : " ", fields, " in $ts")
+    end
+
+    if !isempty(err) && !isempty(type_err)
+        err = "Missing field(s) and unexpected type(s):\n" * err * "\n"
+    end
+    if length(type_err) == 1
+        (field, expected, violation) = only(type_err)
+        err *= "Field `$field` has unexpected type; expected <:$expected, found $violation"
+    elseif length(type_err) > 1
+        err *= "Fields have unexpected type:\n"
+        for (field, expected, violation) in type_err
+            err *= " `$field`: expected <:$expected, found $violation\n"
+        end
+    end
+    throw(ArgumentError(err))
```

Comment:
I'm finding it a bit difficult to follow the error message construction with all of the nested if statements and conditional concatenations...

I wonder if this could be made simpler to read and reason about by being simply additive with each violation found?, e.g.
```julia
err_msg = "Schema violations found:\n"
for err in field_err
    err_msg *= "Missing field: $err\n"
end
for (field, expected, violation) in type_err
    err_msg *= "Incorrect type: `$field` expected `<:$expected`, found $violation \n"
end
throw(ArgumentError(err_msg))
```

(we could even roll this into the loop over `results` above - and remove the needing for constructing the `type_err` or `field_err` vectors - but then we might get a mix of error messages which wouldn't look as neat...)

Reply:
I like this! Thanks, updating

Reply:
done! (slightly modified, see updated example at top of PR :) )